### PR TITLE
ssl: Implement SSL connectivity

### PR DESCRIPTION
**Pull request by Thog**
Depends on #2960.

Supersede #2944.

Adds support for SSL connection using .NET APIs for it.

Compared to the PR it's based on this adds support to non blocking SSL operations and unlink the SSL implementation from the IPC logic.

This PR also more implementation some missing error code/checks that were discovered via a quick reverse of SSL 8.x

### CODE REVIEW

File: Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslConnection.cs [hunk 1]
```diff
@@ -100,19 +324,181 @@ public ResultCode SetSessionCacheMode(ServiceCtx context)
...
+        }
+
         [CommandHipc(22)]
-        // SetOption(b8, nn::ssl::sf::OptionType)
+        // SetOption(nn::ssl::sf::OptionType option, b8 value)
         public ResultCode SetOption(ServiceCtx context)
         {
-            bool       optionEnabled = context.RequestData.ReadBoolean();
-            OptionType optionType    = (OptionType)context.RequestData.ReadUInt32();
+            OptionType option = (OptionType)context.RequestData.ReadUInt32();
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslConnection.cs:364]: Are you sure the order of those parameters is correct here?

Comment (Thog) [reply Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslConnection.cs:364]: libnx follow that order https://github.com/switchbrew/libnx/blob/ea77c1d4cbfd8549afebe840244e493db42ed048/nx/source/services/ssl.c#L334

Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslConnection.cs:364]: That's fo ISslContext though, not ISslConnection.
This is the correct struct https://github.com/switchbrew/libnx/blob/ea77c1d4cbfd8549afebe840244e493db42ed048/nx/source/services/ssl.c#L672

Comment (Thog) [reply Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslConnection.cs:364]: Nice catch 👍

Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslConnection.cs:365]: Padding is still missing no?

File: Ryujinx.HLE/HOS/Services/Ssl/Types/OptionType.cs [hunk 2]
```diff
@@ -5,6 +5,6 @@ enum OptionType : uint
         DoNotCloseSocket,
         GetServerCertChain, // 3.0.0+
         SkipDefaultVerify,  // 5.0.0+
-        EnableAlpn          // 9.0.0+
+        EnableAlpn,         // 9.0.0+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Ssl/Types/OptionType.cs:8]:
```suggestion
        EnableAlpn         // 9.0.0+
```

**Comment by Thog** [comment]
Rebased on top of master

File: Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslConnection.cs [hunk 3]
```diff
@@ -66,32 +185,155 @@ public ResultCode SetIoMode(ServiceCtx context)
...
+            if (_connection == null)
+            {
+                return ResultCode.NoSocket;
+            }
+
+            ResultCode result;
+
+            using (WritableRegion region = context.Memory.GetWritableRegion(context.Request.ReceiveBuff[0].Position, (int)context.Request.ReceiveBuff[0].Size))
+            {
+                // TODO: better error management
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Ssl/SslService/ISslConnection.cs:242]:
```suggestion
                // TODO: Better error management.
```
Same for the other functions.

**Review by gdkchan — APPROVED** [review]
lgtm other than the nits.

**Review by AcK77 — APPROVED** [review]
Just 2 extra empty lines to remove and it’s lgtm!

**Pull request merged**